applications/TrilinosApplication/custom_strategies/convergencecriterias/trilinos_mixed_generic_criteria.h

```diff
@@ -0,0 +1,376 @@
+            }
+            return aux_map;
+        } (rConvergenceVariablesList))
+    {}
+
+    /// Destructor.
+    ~TrilinosMixedGenericCriteria() override {}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /// Compute relative and absoute error.
+    /**
+     * @param rModelPart Reference to the ModelPart containing the fluid problem.
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     * @param A System matrix (unused)
+     * @param Dx Vector of results (variations on nodal variables)
+     * @param b RHS vector (residual)
+     * @return true if convergence is achieved, false otherwise
+     */
+    bool PostCriteria(
+        ModelPart& rModelPart,
+        DofsArrayType& rDofSet,
+        const TSystemMatrixType& A,
+        const TSystemVectorType& Dx,
+        const TSystemVectorType& b) override
+    {
+        // If required, initialize the Epetra vector import
+        if(!mEpetraImportIsInitialized) {
+            InitializeEpetraImport(rDofSet, Dx);
+        }
+
+        // Check if we are solving for something
+        if (TSparseSpace::Size(Dx) != 0) {
+
+            // Initialize
+            std::vector<int> dofs_count(mVariableSize, 0);
+            std::vector<TDataType> solution_norms_vector(mVariableSize, 0.0);
+            std::vector<TDataType> increase_norms_vector(mVariableSize, 0.0);
+
+            int n_dofs = rDofSet.size();
+            const auto& r_data_comm = rModelPart.GetCommunicator().GetDataCommunicator();
+            const int rank = r_data_comm.Rank();
+
+            // Do the local Dx vector import
+            Epetra_Vector local_dx( mpDofImport->TargetMap() );
+            int i_err = local_dx.Import(Dx, *mpDofImport, Insert);
+            KRATOS_ERROR_IF_NOT(i_err == 0) << "Local Dx import failed!" << std::endl;
+
+            // Loop over Dofs
+#pragma omp parallel
+            {
+                // Local thread variables
+                int dof_id;
+                TDataType dof_dx;
+                TDataType dof_value;
+
+                // Local reduction variables
+                std::vector<TDataType> var_solution_norm_reduction(mVariableSize);
+                std::vector<TDataType> var_correction_norm_reduction(mVariableSize);
+                std::vector<int> dofs_counter_reduction(mVariableSize);
+                for (int i = 0; i < mVariableSize; i++) {
+                    var_solution_norm_reduction[i] = 0.0;
+                    var_correction_norm_reduction[i] = 0.0;
+                    dofs_counter_reduction[i] = 0;
+                }
+
+#pragma omp for
+                for (int i = 0; i < n_dofs; i++) {
+                    auto it_dof = rDofSet.begin() + i;
+                    if (it_dof->IsFree() && it_dof->GetSolutionStepValue(PARTITION_INDEX) == rank) {
+                        dof_id = it_dof->EquationId();
```

Comment:
this should be IndexType which might be larger than a int can contain

Reply:
OK. Thanks!

Reply:
Done.

Reply:
Error. The `Epetra_BlockMap` does not resolve `std::size_t`  as type for the `dof_id`. There are two candidate functions, one with `int` and one with `long long`. As we rarely use the `long long` type, I'd prefer to keep it as it is.

Reply:
ok